helix-view/src/editor.rs

```diff
@@ -2154,6 +2160,92 @@ impl Editor {
             current_view.id
         }
     }
+
+    pub fn add_diagnostics(
+        &mut self,
+        diagnostics: Vec<(lsp::Diagnostic, LanguageServerId)>,
+        server_id: LanguageServerId,
+        uri: helix_core::Uri,
+        document_version: Option<i32>,
+        result_id: Option<String>,
+    ) {
+        let doc = self.documents.values_mut()
+                            .find(|doc| doc.uri().is_some_and(|u| u == uri))
+                            .filter(|doc| {
+                                if let Some(version) = document_version{
+                                    if version != doc.version() {
+                                        log::info!("Version ({version}) is out of date for {uri:?} (expected ({}), dropping PublishDiagnostic notification", doc.version());
+                                        return false;
+                                    }
+                                }
+                                true
+                            });
+
+        if let Some(doc) = doc {
```

Comment:
While we're moving this block let's take the opportunity to refactor some of the code here:

```rust
let Some(doc) = self
    .documents
    .values_mut()
    .find(|doc| doc.uri().is_some_and(|u| u == uri))
else {
    return;
};
if let Some(version) = document_version {
    if version != doc.version() {
        log::info!("Version ({version}) is out of date for {uri:?} (expected ({}), dropping PublishDiagnostic notification", doc.version());
        return;
    }
}
```

That eliminates the extra nesting from `if let Some(doc) = doc` and IMO the `find` + `filter` is a little hard to follow in its current form.

Reply:
Agree. This is much better!